10x faster journalctl: cache timestamps

ktsaou opened the pull request:
Extracted and improved from: https://github.com/systemd/systemd/pull/29261

## The Problem

As demonstrated in PR https://github.com/systemd/systemd/pull/29261, while querying journal files the library is trying to repeatedly extract the timestamps of the last message in each journal file, slowing down significantly the query.

Current main (190.3 seconds):

![image](https://github.com/systemd/systemd/assets/2662304/1ed35067-d8c5-49b4-966e-d12954afe6d7)


## The Solution

- [x] Added option `SD_JOURNAL_FAST_QUERY`.
- [x] When set, the latest timestamps of each journal file are cached for the life time of the query.

This PR (18.3 seconds, about 10.5x faster than current main branch):

![image](https://github.com/systemd/systemd/assets/2662304/41845651-2c44-4750-9567-851c33f2ca6d)









<!-- devel-freezer = {"comment-id":"1843061980","freezing-tag":"v255-rc4"} -->

yuwata commented on src/systemd/sd-journal.h:
```diff
@@ -72,6 +72,7 @@ enum {
         SD_JOURNAL_ALL_NAMESPACES            = 1 << 5, /* Show all namespaces, not just the default or specified one */
         SD_JOURNAL_INCLUDE_DEFAULT_NAMESPACE = 1 << 6, /* Show default namespace in addition to specified one */
         SD_JOURNAL_TAKE_DIRECTORY_FD         = 1 << 7, /* sd_journal_open_directory_fd() will take ownership of the provided file descriptor. */
+        SD_JOURNAL_FAST_QUERY                = 1 << 8, /* cache fstat() calls */
```
Please make the name more descriptive and specific, e.g. `SD_JOURNAL_CALL_FSTAT_ONLY_ONCE` or so.

ktsaou replied:
A more descriptive term would be `SD_JOURNAL_CACHE_FILES_NEWEST_TIMESTAMPS`. The reason is the caching is applied way before `fstat()` is called, to cache the newest timestamps of journal files. There is no `fstat()` caching on this PR.

What would you prefer?

yuwata replied:
> `SD_JOURNAL_CACHE_FILES_NEWEST_TIMESTAMPS`

Sounds good, though I have not followed the code in detail yet.

yuwata commented on src/journal/journalctl.c:
```diff
@@ -2280,20 +2280,20 @@ static int run(int argc, char *argv[]) {
 
         if (arg_directory)
-                r = sd_journal_open_directory(&j, arg_directory, arg_journal_type);
+                r = sd_journal_open_directory(&j, arg_directory, arg_journal_type | SD_JOURNAL_FAST_QUERY);
```
I think here we should not set the flag unconditionally, as the specified directory may be /var/build/journal/remote.

ktsaou replied:
What is `/var/build/journal/remote` and why it should be different during querying by `journalctl`?
I don't see any references in the code to understand the issue.
Can you help me?

yuwata replied:
sorry, I have no idea why I include 'build'...
The directory I wanted to say is `/var/log/journal/remote`.

yuwata replied:
Hmm. I completely forget why I thought so. Feel free to ignore.
I will re-review later.

yuwata commented on src/journal/journalctl.c:
```diff
@@ -2280,20 +2280,20 @@ static int run(int argc, char *argv[]) {
+                r = sd_journal_open_directory(&j, arg_directory, arg_journal_type | SD_JOURNAL_FAST_QUERY);
         else if (arg_root)
-                r = sd_journal_open_directory(&j, arg_root, arg_journal_type | SD_JOURNAL_OS_ROOT);
+                r = sd_journal_open_directory(&j, arg_root, arg_journal_type | SD_JOURNAL_OS_ROOT | SD_JOURNAL_FAST_QUERY);
```
Here also.

yuwata reviewed (changes requested):
I cannot follow the change. Please explain the change in the commit message, and squash the two commits.

ktsaou commented:
> I cannot follow the change. Please explain the change in the commit message, and squash the two commits.

`journalctl` may open journals in multiple ways. In the original commit, I missed that. In the later commit I added the flag on all calls.

I will squash.

lilydjwg commented:
I applied this patch to 254.6. It no longer seemed to work (still very slow).

yuwata commented:
I cannot observe performance regression caused by fstat().
More simplified version is in #30209.

lilydjwg commented:
> I applied this patch to 254.6. It no longer seemed to work (still very slow).

I missed that a new flag should be specified. After adding the flag, it still works.

ktsaou commented:
Updated to current main.

Run the test again:

This is the main branch:
![image](https://github.com/systemd/systemd/assets/2662304/e65f8fde-9a17-47f5-8e05-1238040ae146)

And this PR:
![image](https://github.com/systemd/systemd/assets/2662304/5bcaae0e-0dd0-49d4-a2e8-ed0aab18578f)

As you can see, 26 seconds vs 3 seconds.

The test envirionment:

`journalctl` compiled with this:
rm -rf build; meson --buildtype=debugoptimized build . && ninja -C build version.h && ninja -C build libsystemd && ninja -C build journalctl
```sh

`journalctl` invoked like this:
./build/journalctl --no-pager --since 2023-11-25
```sh

For the effects of this PR to be noticable you need to have multiple journal files participating in the query. The library refreshes the timestamps every time it checks if it should switch files. So, if you are testing queries that match a single file, this PR has no effect. Only when you have a lot of files in a single query you are going to see its effects.

```
```

ktsaou commented:
@yuwata I renamed the variable and squashed the PR.
In my tests, this is still valid and has a huge impact on multi-file journal queries.
It would be ideal to be merged asap.

lilydjwg commented:
> So, if you are testing queries that match a single file, this PR has no effect. Only when you have a lot of files in a single query you are going to see its effects.

I have 101 files and 821.3M in size now, and it's still growing.

yuwata commented:
So, is it OK to close this ? I hope #30209 can replace this.

ktsaou commented:
> So, is it OK to close this ? I hope #30209 can replace this.

Yes, it is ok. Let's continue there.
